File: 2021/11/12/ticTacToe.js
/**
 * https://leetcode.com/problems/find-winner-on-a-tic-tac-toe-game/
 * @author Andrew Jarombek
 * @since 11/12/2021
 */

const tictactoe = function(moves) {
    const board = [['', '', ''], ['', '', ''], ['', '', '']]

    for (let i = 0; i < moves.length; i++) {
        const move = moves[i];
        board[move[0]][move[1]] = i % 2 === 0 ? 'A' : 'B';

        for (let j = 0; j < board.length; j++) {
            if (board[j][0] && board[j][0] === board[j][1] && board[j][1] === board[j][2]) {
                return board[j][0];
            }

            if (board[0][j] && board[0][j] === board[1][j] && board[1][j] === board[2][j]) {
                return board[0][j];
            }
        }

        if (board[0][0] && board[0][0] === board[1][1] && board[1][1] === board[2][2]) {
            return board[0][0];
        }

        if (board[0][2] && board[0][2] === board[1][1] && board[1][1] === board[2][0]) {
            return board[0][2];
        }
    }

    for (const row of board) {
        for (const cell of row) {
            if (cell === '') {
                return 'Pending';
            }
        }
    }

    return 'Draw';
};

File: 2022/08/02/tictactoe.py
"""
https://leetcode.com/problems/find-winner-on-a-tic-tac-toe-game/
Author: Andrew Jarombek
Date: 8/2/2022
"""

from typing import List


def tictactoe(moves: List[List[int]]) -> str:
    board = [['0', '1', '2'], ['3', '4', '5'], ['6', '7', '8']]

    for index, move in enumerate(moves):
        board[move[0]][move[1]] = 'A' if index % 2 == 0 else 'B'

    for i in range(3):
        if board[i][0] == board[i][1] == board[i][2]:
            return board[i][0]

    for i in range(3):
        if board[0][i] == board[1][i] == board[2][i]:
            return board[0][i]

    if board[0][0] == board[1][1] == board[2][2]:
        return board[0][0]

    if board[2][0] == board[1][1] == board[0][2]:
        return board[1][1]

    return "Pending" if len(moves) < 9 else "Draw"


if __name__ == '__main__':
    assert tictactoe([[0, 0], [2, 0], [1, 1], [2, 1], [2, 2]]) == "A"
    assert tictactoe([[1, 2], [2, 1], [0, 0], [0, 1], [2, 0], [0, 2], [2, 2], [1, 1]]) == "B"
    assert tictactoe([[0, 0], [1, 1], [2, 0], [1, 0], [1, 2], [2, 1], [0, 1], [0, 2], [2, 2]]) == "Draw"
    assert tictactoe([[0, 0], [1, 1], [2, 0], [1, 0], [1, 2], [2, 1], [0, 1], [0, 2]]) == "Pending"
